### PR TITLE
Лопатин Илья. Задача 2. Вариант 15. Ленточная горизонтальная схема.

### PR DESCRIPTION
**Pull request by ascannel**
- seq
1. using an outer loop we go through the rows of the matrix
2. we go through the columns with an inner loop
3. inside loops, we calculate the element of the resulting vector by summing the multiplied rows of the matrix by the vector

- mpi
1. send matrix and vector sizes
2. determine strip size for each process
3. allocating memory for the local part of the matrix and local vector for each process
4. sending local vector to processes
5. sending parts of the matrix to processes
6. calculate the local result
7. collecting results

### CODE REVIEW

File: tasks/seq/lopatin_i_strip_horizontal_scheme/src/stripHorizontalSchemeSeq.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#include "seq/lopatin_i_strip_horizontal_scheme/include/stripHorizontalSchemeHeaderSeq.hpp"
+
+namespace lopatin_i_strip_horizontal_scheme_seq {
+
+bool lopatin_i_strip_horizontal_scheme_seq::TestTaskSequential::validation() {
```
Comment (Ivan-Lysov) [inline tasks/seq/lopatin_i_strip_horizontal_scheme/src/stripHorizontalSchemeSeq.cpp:5]: You have already defined the namespace above, there is no need to define it in the function

File: tasks/seq/lopatin_i_strip_horizontal_scheme/perf_tests/stripHorizontalSchemePerfTestsSeq.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+TEST(lopatin_i_strip_horizontal_scheme_seq, test_pipeline_run) {
+  std::vector<int> inputMatrix = testMatrix;
+  std::vector<int> inputVector = testVector;
+  std::vector<int> resultVector(2160, 0);
+
+  std::shared_ptr<ppc::core::TaskData> taskDataSequential = std::make_shared<ppc::core::TaskData>();
+  taskDataSequential->inputs.emplace_back(reinterpret_cast<uint8_t *>(inputMatrix.data()));
+  taskDataSequential->inputs_count.emplace_back(6);
+  taskDataSequential->inputs_count.emplace_back(4);
```
Comment (Ivan-Lysov) [inline tasks/seq/lopatin_i_strip_horizontal_scheme/perf_tests/stripHorizontalSchemePerfTestsSeq.cpp:38]: You have set the wrong dimensions of the matrix. You generate a matrix of the same size and the test uses a different matrix

File: tasks/seq/lopatin_i_strip_horizontal_scheme/src/stripHorizontalSchemeSeq.cpp (hunk 1 above)
Comment (ascannel) [reply tasks/seq/lopatin_i_strip_horizontal_scheme/src/stripHorizontalSchemeSeq.cpp]: thx, fixed

File: tasks/seq/lopatin_i_strip_horizontal_scheme/perf_tests/stripHorizontalSchemePerfTestsSeq.cpp (hunk 2 above)
Comment (ascannel) [reply tasks/seq/lopatin_i_strip_horizontal_scheme/perf_tests/stripHorizontalSchemePerfTestsSeq.cpp]: thx, fixed

**Review by Ivan-Lysov — APPROVED** [review]

File: tasks/mpi/lopatin_i_strip_horizontal_scheme/src/stripHorizontalSchemeMPI.cpp [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  return outputVector;
+}
+
+std::vector<int> generateMatrix(int sizeX, int sizeY) {
+  std::random_device dev;
+  std::mt19937 gen(dev());
+  int matrixSize = sizeX * sizeY;
+  std::vector<int> outputMatrix(matrixSize);
+  for (int i = 0; i < matrixSize; i++) {
+    outputMatrix[i] = (gen() % 200);
```
Comment (KolodkinGrigorii) [inline tasks/mpi/lopatin_i_strip_horizontal_scheme/src/stripHorizontalSchemeMPI.cpp:21]: add negative values

Comment (ascannel) [reply tasks/mpi/lopatin_i_strip_horizontal_scheme/src/stripHorizontalSchemeMPI.cpp]: thank you, fixed

**Review by KolodkinGrigorii — APPROVED** [review]

File: tasks/mpi/lopatin_i_strip_horizontal_scheme/src/stripHorizontalSchemeMPI.cpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+std::vector<int> generateMatrix(int sizeX, int sizeY) {
+  std::random_device dev;
+  std::mt19937 gen(dev());
+  int matrixSize = sizeX * sizeY;
+  std::vector<int> outputMatrix(matrixSize);
+  for (int i = 0; i < matrixSize; i++) {
+    outputMatrix[i] = (gen() % 200) - 99;
+  }
+  return outputMatrix;
+}
```
Comment (allnes) [inline tasks/mpi/lopatin_i_strip_horizontal_scheme/src/stripHorizontalSchemeMPI.cpp:24]: please move it in test files

File: tasks/mpi/lopatin_i_strip_horizontal_scheme/func_tests/stripHorizontalSchemeFuncTestsMPI.cpp [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+    ASSERT_TRUE(testTaskSequential.validation());
+    testTaskSequential.pre_processing();
+    testTaskSequential.run();
+    testTaskSequential.post_processing();
+
+    ASSERT_EQ(resultVector, referenceResultVector);
+  }
+}
+
+TEST(lopatin_i_strip_horizontal_scheme_mpi, test_2560x1440_matrix) {
```
[the remark this answers is not in the document]
Comment (ascannel) [reply tasks/mpi/lopatin_i_strip_horizontal_scheme/func_tests/stripHorizontalSchemeFuncTestsMPI.cpp:277]: maybe you meant the case when the number of rows in the matrix is ​​not divisible by the number of processes?

[the remark this answers is not in the document]
Comment (ascannel) [reply tasks/mpi/lopatin_i_strip_horizontal_scheme/func_tests/stripHorizontalSchemeFuncTestsMPI.cpp:277]: or, if i understood you correctly, the program should be able to handle cases when the number of input data (in this case, the number of rows in the matrix) is less than the number of processes on which we are running?

File: tasks/mpi/lopatin_i_strip_horizontal_scheme/src/stripHorizontalSchemeMPI.cpp (hunk 4 above)
Comment (ascannel) [reply tasks/mpi/lopatin_i_strip_horizontal_scheme/src/stripHorizontalSchemeMPI.cpp]: moved

File: tasks/mpi/lopatin_i_strip_horizontal_scheme/func_tests/stripHorizontalSchemeFuncTestsMPI.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ascannel) [reply tasks/mpi/lopatin_i_strip_horizontal_scheme/func_tests/stripHorizontalSchemeFuncTestsMPI.cpp:344]: ok, i've figured out both, added

**Review by allnes — APPROVED** [review]

**Review by aobolensk — APPROVED** [review]

**Pull request merged**